### ISSUE TITLE
Questions Regarding Missing Code for pusht_stablizing and Discrete Actions in Robomimic Datasets

### ISSUE THREAD

**Issue by ischencheng** [siddancha/streaming-flow-policy#13]
Hi authors, thank you for the great work and for open-sourcing the code.

While going through the repository, I noticed that I couldn’t find the code related to pusht_stablizing mentioned in the paper. Do you have plans to release that code? If the “stabilizing” mechanism is simply adding noise to the target velocity, could you clarify how the noise scale k is selected? With this approach, I’ve found it difficult to reproduce your reported results on Push-T.

In addition, I’m also trying to reproduce your experiments on Robomimic datasets (e.g., lift) but couldn’t find relevant code in the repo. Could you please clarify how you handled discrete variables, especially the gripper action? Any information on preprocessing or model adaptation for these discrete components would be helpful.

**Comment by siddancha** [comment]
Hi @ischencheng, thanks for your interest in our work! As you've realized, we've currently only released source code to reproduce our notebooks containing the toy 1D example. Code release for Push-T and Robomimic is currently WIP.

> While going through the repository, I noticed that I couldn’t find the code related to pusht_stablizing mentioned in the paper. Do you have plans to release that code?

Yes, we're planning to release a self-contained Colab notebook for training and evaluating Streaming Flow Policy (SFP) on the Push-T environment, and comparing it to diffusion policy. We will do this by the end of this month (because some of the authors, including myself, are currently on vacation 😅). Hope you're okay waiting till then?

> If the “stabilizing” mechanism is simply adding noise to the target velocity, could you clarify how the noise scale k is selected?

To clarify, noise is added to the state (input), not the target (velocity). Specifically, if $\xi: [0, 1] \rightarrow \mathcal{A}$ denotes an action chunk from the training set, the sampling procedure is:

$$
\begin{align*}
\ \ \ \ \ \ t &\sim \mathrm{Uniform}([0, 1])\\
\ \ \ \ \ \ x &= \xi(t) + \sigma_0 \exp(-kt) \epsilon ~~~~;~~ \epsilon \sim \mathcal{N}(0, 1)~~~\text{(noise added in this step)}\\
\ \ \ \ \ \ v &=\frac{d}{dt}{\xi}(t) - k (x - \xi(t))~~~\text{(no noise added in this step)}\\
\text{Loss} &= \Vert v_\theta(x, t) - v \Vert^2_2
\end{align*}
$$

$k$ is a hyperparameter that is tuned. We found that the method is not very sensitive to $k$ as long as $k > 0$. The following hyperparameters worked well for us on Push-T: $\sigma_0 = 0.4$, $k = 15$ and `num_epochs = 100`. Note that at test time, we set $\sigma_0 = 0$ and run the policy deterministically. Again, we will release the code (notebook) soon which has all implementation details.

> For Robomimic, could you please clarify how you handled discrete variables, especially the gripper action?

That's a great question. Discrete action spaces need to be handled specially, since SFP technically requires action trajectories to be continuous and differentiable with respect to time. We converted binary open/close gripper actions $\lbrace 0, 1\rbrace$ to the continuous interval $[0, 1]$ (the underlying gripper open/close position is in fact continuous). Furthermore, simply differentiating the gripper trajectories to target compute velocities is not a good idea, because then target velocities are almost always zero except at a few timesteps when the gripper transitions from open to close or vice-versa. So gripper velocity is modeled as a large value $v_G \approx 1/\Delta t$ when $a = 1$ and $v_G \approx -1/\Delta t$ when $a = 0$, where $\Delta t$ is the ODE integration timestep. Furthermore, during ODE integration as test time, the gripper action is enforced to lie in $[0, 1]$ — if the value overflows beyond $[0, 1]$, it is truncated to 0 or 1. 

Hope that helps! All the implementation details will be released soon. Thanks for your patience! Feel free to email sancha@mit.edu if you have more questions.
